keps/sig-network/2136-np-version-status/README.md

```diff
@@ -0,0 +1,858 @@
+implementations should create an "empty" `status` right away if it's
+going to take them a while to determine the final `status`.
+
+#### Distributed and Delegating NetworkPolicy Implementations
+
+Some NetworkPolicy implementations do not operate from a single
+controller, but instead operate independently on every node (much like
+how kube-proxy implements Service rules). In this case it is not clear
+who would be responsible for updating `NetworkPolicyStatus`, and it may
+not be easy for any part of the system to know with certainty that a
+given NetworkPolicy was fully in effect across the cluster.
+
+In other cases, a plugin may implement NetworkPolicy in terms of some
+other lower-level policy engine (implemented by a cloud, perhaps), and
+may work by simply translating NetworkPolicy rules into rules for this
+other engine. As in the distributed case, it may not be possible for
+such a NetworkPolicy implementation to know for sure when a policy was
+fully in effect.
+
+### Risks and Mitigations
+
+## Design Details
+
+### API
+
+Add a new field to `networkingv1.NetworkPolicySpec`:
+
+```
+type NetworkPolicySpec struct {
+	...
+
+	// minVersion indicates the Kubernetes release version corresponding to
+	// the NetworkPolicy functionality required by this policy. If this is
+	// explicitly specified by the user, the apiserver will require that the policy
+	// conforms to the specified version. If it is not specified, the apiserver
+	// will fill in the correct minVersion based on the features used by the policy.
+	// +optional
+	MinVersion NetworkPolicyVersion `json:"minVersion,omitempty" protobuf:"bytes,5,name=minVersion"`
+
+	...
+}
+
+type NetworkPolicyVersion string
+
+const (
+	// Kubernetes 1.3 introduced the NetworkPolicy v1beta1 API, supporting
+	// ingress-only policies containing podSelectors and namespaceSelectors,
+	// and TCP and UDP ports.
+	NetworkPolicyVersion1_3 NetworkPolicyVersion = "1.3"
+
+	// Kubernetes 1.8 brought the NetworkPolicy API to v1, adding egress policies
+	// and ipBlocks.
+	NetworkPolicyVersion1_8 NetworkPolicyVersion = "1.8"
+
+	// Kubernetes 1.9 added alpha IPv6 support. Plugins implementing this version are
+	// assumed to recognize IPv6 addresses, even if they don't implement them.
+	NetworkPolicyVersion1_9 NetworkPolicyVersion = "1.9"
+
+	// Kubernetes 1.11 added support for specifying podSelector and namespaceSelector
+	// together in a peer.
+	NetworkPolicyVersion1_11 NetworkPolicyVersion = "1.11"
+
+	// Kubernetes 1.12 added alpha SCTP support. Plugins implementing this version are
+	// assumed to recognize `protocol: SCTP`, even if they don't implement it.
+	NetworkPolicyVersion1_12 NetworkPolicyVersion = "1.12"
+)
+
+```
+
+And add a new `Status NetworkPolicyStatus` field to `NetworkPolicy`, defined as:
+
+```
+// NetworkPolicyStatus contains information about the processing of a NetworkPolicy
+type NetworkPolicyStatus struct {
```

Comment:
I am tentatively OK with adding conditions, but we need to define what each condition really means.  What happens when Enforcing is true, then I update the policy?  Do we auto-reset it to false, or do we leave it stale?  What happens if one node out of 5000 is out to lunch, and is not implementing the policy?  What if I have different versions of the agent on different subsets of nodes?

Reply:
> What happens when Enforcing is true, then I update the policy? Do we auto-reset it to false, or do we leave it stale?

Hm... yeah, I was only thinking about creation time. What happened with pod ReadinessGates? I feel like people agreed that NetworkPolicy programming should block pod readiness at creation time, but pods should _not_ become unready later on just because new policies were added.

Even if we say Enforcing doesn't track updates, if you really cared about that case you could know when it was ready by creating a new policy rather than updating the old one...

> What happens if one node out of 5000 is out to lunch, and is not implementing the policy? What if I have different versions of the agent on different subsets of nodes?

I feel like those questions must apply to the sorts of things people wanted Pod ReadinessGates for too. Did we come up with answers there?

Reply:
We decided that readiness was exclusively about the initial readiness.  But that's driven by small-scale signals.  E.g. "is my LB programmed" not "is every single node in my 5K cluster reporting ready"